Compilation error with pythonize 0.21 and pyo3 0.21.1

I've tried updating a private project to pyo3 0.21.1 and pythonize 0.21.0, but it fails to compile with the following issue:

```
error[E0599]: no method named `to_str` found for reference `&pyo3::Bound<'_, PyUnicode>` in the current scope
   --> /home/dobo/.cargo/registry/src/index.crates.io-6f17d22bba15001f/pythonize-0.21.0/src/de.rs:133:52
    |
133 |         let s = self.input.downcast::<PyString>()?.to_str()?;
    |                                                    ^^^^^^ help: there is a method with a similar name: `into_ptr`

error[E0599]: no method named `to_str` found for reference `&pyo3::Bound<'_, PyUnicode>` in the current scope
   --> /home/dobo/.cargo/registry/src/index.crates.io-6f17d22bba15001f/pythonize-0.21.0/src/de.rs:156:29
    |
156 |         visitor.visit_str(s.to_str()?)
    |                             ^^^^^^ help: there is a method with a similar name: `into_ptr`

error[E0599]: no method named `to_str` found for reference `&pyo3::Bound<'_, PyUnicode>` in the current scope
   --> /home/dobo/.cargo/registry/src/index.crates.io-6f17d22bba15001f/pythonize-0.21.0/src/de.rs:286:34
    |
286 |             visitor.visit_enum(s.to_str()?.into_deserializer())
    |                                  ^^^^^^ help: there is a method with a similar name: `into_ptr`

error[E0599]: no method named `to_str` found for reference `&pyo3::Bound<'_, PyUnicode>` in the current scope
   --> /home/dobo/.cargo/registry/src/index.crates.io-6f17d22bba15001f/pythonize-0.21.0/src/de.rs:300:29
    |
300 |         visitor.visit_str(s.to_str()?)
    |                             ^^^^^^ help: there is a method with a similar name: `into_ptr`

error[E0599]: no method named `to_str` found for struct `pyo3::Bound<'py, PyUnicode>` in the current scope
   --> /home/dobo/.cargo/registry/src/index.crates.io-6f17d22bba15001f/pythonize-0.21.0/src/de.rs:409:26
    |
409 |             self.variant.to_str()?.into_deserializer();
    |                          ^^^^^^ help: there is a method with a similar name: `into_ptr`
``` 

Any idea what could be going wrong here?

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -68,6 +68,9 @@ jobs:
       - name: Test
         run: cargo test --verbose
 
+      - name: Test (abi3)
+        run: cargo test --verbose --features pyo3/abi3-py37
+
     env:
       RUST_BACKTRACE: 1
 

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,3 +1,7 @@
+## 0.21.1 - 2024-04-02
+
+- Fix compile error when using PyO3 `abi3` feature targeting a minimum version below 3.10
+
 ## 0.21.0 - 2024-04-01
 
 - Bump edition to 2021

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "pythonize"
-version = "0.21.0"
+version = "0.21.1"
 authors = ["David Hewitt <1939362+davidhewitt@users.noreply.github.com>"]
 edition = "2021"
 rust-version = "1.56"

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -17,7 +17,7 @@ pyo3 = { version = "0.21.0", default-features = false }
 
 [dev-dependencies]
 serde = { version = "1.0", default-features = false, features = ["derive"] }
-pyo3 = { version = "0.21.0", default-features = false, features = ["auto-initialize", "macros"] }
+pyo3 = { version = "0.21.1", default-features = false, features = ["auto-initialize", "macros"] }
 serde_json = "1.0"
 maplit = "1.0.2"
 serde_path_to_error = "0.1.15"

--- a/src/de.rs
+++ b/src/de.rs
@@ -130,7 +130,7 @@ impl<'a, 'py, 'de> de::Deserializer<'de> for &'a mut Depythonizer<'py> {
     where
         V: de::Visitor<'de>,
     {
-        let s = self.input.downcast::<PyString>()?.to_str()?;
+        let s = self.input.downcast::<PyString>()?.to_cow()?;
         if s.len() != 1 {
             return Err(PythonizeError::invalid_length_char());
         }
@@ -153,7 +153,7 @@ impl<'a, 'py, 'de> de::Deserializer<'de> for &'a mut Depythonizer<'py> {
         V: de::Visitor<'de>,
     {
         let s = self.input.downcast::<PyString>()?;
-        visitor.visit_str(s.to_str()?)
+        visitor.visit_str(&s.to_cow()?)
     }
 
     fn deserialize_string<V>(self, visitor: V) -> Result<V::Value>
@@ -283,7 +283,7 @@ impl<'a, 'py, 'de> de::Deserializer<'de> for &'a mut Depythonizer<'py> {
             let mut de = Depythonizer::from_object_bound(value);
             visitor.visit_enum(PyEnumAccess::new(&mut de, variant))
         } else if let Ok(s) = item.downcast::<PyString>() {
-            visitor.visit_enum(s.to_str()?.into_deserializer())
+            visitor.visit_enum(s.to_cow()?.into_deserializer())
         } else {
             Err(PythonizeError::invalid_enum_type())
         }
@@ -297,7 +297,7 @@ impl<'a, 'py, 'de> de::Deserializer<'de> for &'a mut Depythonizer<'py> {
             .input
             .downcast::<PyString>()
             .map_err(|_| PythonizeError::dict_key_not_string())?;
-        visitor.visit_str(s.to_str()?)
+        visitor.visit_str(&s.to_cow()?)
     }
 
     fn deserialize_ignored_any<V>(self, visitor: V) -> Result<V::Value>
@@ -405,8 +405,8 @@ impl<'a, 'py, 'de> de::EnumAccess<'de> for PyEnumAccess<'a, 'py> {
     where
         V: de::DeserializeSeed<'de>,
     {
-        let de: de::value::StrDeserializer<'_, PythonizeError> =
-            self.variant.to_str()?.into_deserializer();
+        let cow = self.variant.to_cow()?;
+        let de: de::value::StrDeserializer<'_, PythonizeError> = cow.as_ref().into_deserializer();
         let val = seed.deserialize(de)?;
         Ok((val, self))
     }

--- a/src/ser.rs
+++ b/src/ser.rs
@@ -478,6 +478,7 @@ mod test {
     use super::pythonize;
     use maplit::hashmap;
     use pyo3::prelude::*;
+    use pyo3::pybacked::PyBackedStr;
     use pyo3::types::PyDict;
     use serde::Serialize;
 
@@ -497,7 +498,7 @@ mod test {
                 Some(&locals),
             )?;
             let result = locals.get_item("result")?.unwrap();
-            let result = result.extract::<&str>()?;
+            let result = result.extract::<PyBackedStr>()?;
 
             assert_eq!(result, expected);
             assert_eq!(serde_json::to_string(&src).unwrap(), expected);

